File: kafka/consumer.py
import json
import prometheus_client
from kafka import KafkaConsumer
from prometheus_client import start_http_server
import os 

PROMETHEUS_POST_NUM = os.environ['PROMETHEUS_PORT']
KAFKA_TOPIC = os.environ['KAFKA_TOPIC']
KAFKA_HOST = os.environ['KAFKA_HOST']
KAFKA_PORT = os.environ['KAFKA_PORT']

# Welford's algorithm for calculating mean and variance without saving all data
class Welford:
    def __init__(self):
        self.mean_x1 = 0
        self.mean_x2 = 0
        self.var_x1 = 0
        self.var_x2 = 0
        self.n = 0

    def update(self, x1, x2):
        self.n += 1
        delta_x1 = x1 - self.mean_x1
        self.mean_x1 += delta_x1 / self.n
        self.var_x1 += delta_x1 * (x1 - self.mean_x1)

        delta_x2 = x2 - self.mean_x2
        self.mean_x2 += delta_x2 / self.n
        self.var_x2 += delta_x2 * (x2 - self.mean_x2)

    def get_mean_and_variance(self):
        return self.mean_x1, self.var_x1 / (self.n - 1) if self.n > 1 else 0, self.mean_x2, self.var_x2 / (self.n - 1) if self.n > 1 else 0

# Prometheus Metrics
mean_x1_metric = prometheus_client.Gauge('train_input_mean_x1', 'Mean of x1 from training input data')
var_x1_metric = prometheus_client.Gauge('train_input_var_x1', 'Variance of x1 of training input data')
mean_x2_metric = prometheus_client.Gauge('train_input_mean_x2', 'Mean of x2 from training input data')
var_x2_metric = prometheus_client.Gauge('train_input_var_x2', 'Variance of x2 of training input data')

# Kafka consumer setup
def create_consumer():
    consumer = KafkaConsumer(
        KAFKA_TOPIC,
        bootstrap_servers=f'{KAFKA_HOST:KAFKA_PORT}',
        value_deserializer=lambda x: json.loads(x.decode('utf-8'))
    )
    return consumer

def main():
    welford = Welford()
    consumer = create_consumer()
    
    start_http_server(PROMETHEUS_POST_NUM)
    
    for message in consumer:
        data = message.value
        x1 = data.get('x1')
        x2 = data.get('x2')
        
        if x1 is not None and x2 is not None:
            welford.update(x1, x2)
            mean_x1, var_x1, mean_x2, var_x2 = welford.get_mean_and_variance()

            mean_x1_metric.set(mean_x1)
            var_x1_metric.set(var_x1)
            mean_x2_metric.set(mean_x2)
            var_x2_metric.set(var_x2)

if __name__ == '__main__':
    main()

File: kafka/consumer.js
const { Kafka } = require('kafkajs');
const prometheus = require('prom-client');
const http = require('http');
const process = require('process');

// Prometheus metrics
const meanX1Metric = new prometheus.Gauge({
  name: 'train_input_mean_x1',
  help: 'Mean of x1 from training input data'
});

const varX1Metric = new prometheus.Gauge({
  name: 'train_input_var_x1',
  help: 'Variance of x1 of training input data'
});

const meanX2Metric = new prometheus.Gauge({
  name: 'train_input_mean_x2',
  help: 'Mean of x2 from training input data'
});

const varX2Metric = new prometheus.Gauge({
  name: 'train_input_var_x2',
  help: 'Variance of x2 of training input data'
});

// Welford's algorithm for calculating mean and variance
class Welford {
  constructor() {
    this.meanX1 = 0;
    this.meanX2 = 0;
    this.varX1 = 0;
    this.varX2 = 0;
    this.n = 0;
  }

  update(x1, x2) {
    this.n += 1;
    
    const deltaX1 = x1 - this.meanX1;
    this.meanX1 += deltaX1 / this.n;
    this.varX1 += deltaX1 * (x1 - this.meanX1);

    const deltaX2 = x2 - this.meanX2;
    this.meanX2 += deltaX2 / this.n;
    this.varX2 += deltaX2 * (x2 - this.meanX2);
  }

  getMeanAndVariance() {
    return [
      this.meanX1, 
      this.n > 1 ? this.varX1 / (this.n - 1) : 0,
      this.meanX2, 
      this.n > 1 ? this.varX2 / (this.n - 1) : 0
    ];
  }
}

// Kafka consumer setup
const createConsumer = () => {
  const kafka = new Kafka({
    clientId: 'consumer',
    brokers: [`${process.env.KAFKA_HOST}:${process.env.KAFKA_PORT}`]
  });

  const consumer = kafka.consumer({ groupId: 'trainingdata' });
  return consumer;
};

// Start HTTP server for Prometheus scraping
const startHttpServer = () => {
  const port = process.env.PROMETHEUS_PORT;
  http.createServer((req, res) => {
    if (req.url === '/metrics') {
      res.setHeader('Content-Type', prometheus.register.contentType);
      res.end(prometheus.register.metrics());
    } else {
      res.statusCode = 404;
      res.end();
    }
  }).listen(port, () => {
    console.log(`Prometheus server listening on port ${port}`);
  });
};

// Main function
const main = async () => {
  const welford = new Welford();
  const consumer = createConsumer();
  
  // Start Prometheus server
  startHttpServer();

  // Connect Kafka consumer and subscribe to topic
  await consumer.connect();
  await consumer.subscribe({ topic: process.env.KAFKA_TOPIC, fromBeginning: true });

  // Consume messages from Kafka
  await consumer.run({
    eachMessage: async ({ message }) => {
      const data = JSON.parse(message.value.toString());
      const x1 = data.x1;
      const x2 = data.x2;

      if (x1 !== undefined && x2 !== undefined) {
        welford.update(x1, x2);
        const [meanX1, varX1, meanX2, varX2] = welford.getMeanAndVariance();

        // Update Prometheus metrics
        meanX1Metric.set(meanX1);
        varX1Metric.set(varX1);
        meanX2Metric.set(meanX2);
        varX2Metric.set(varX2);
      }
    }
  });
};

main().catch(console.error);
